File: web/src/lib/crypto/key-backup.test.ts
// @vitest-environment node
// Mirrors android/core/crypto KeyBackupTest.kt over the v1 blob format.

import { expect, it } from "vitest";
import { randomBytes, toHex } from "./bytes";
import {
  KeyBackupError,
  openKeyBackup,
  RecoveryCode,
  RecoveryCodeLengthError,
  sealKeyBackup,
} from "./key-backup";

it("seal and open round-trip", async () => {
  const seed = randomBytes(32);
  const code = RecoveryCode.generate();
  const blob = await sealKeyBackup(seed, code);
  expect(toHex(await openKeyBackup(blob, code))).toBe(toHex(seed));
});

it("a wrong code does not open", async () => {
  const blob = await sealKeyBackup(randomBytes(32), RecoveryCode.generate());
  await expect(openKeyBackup(blob, RecoveryCode.generate())).rejects.toThrow(
    new KeyBackupError("the blob does not open under this code"),
  );
});

it("a tampered header does not open", async () => {
  const code = RecoveryCode.generate();
  const blob = await sealKeyBackup(randomBytes(32), code);
  blob[5] ^= 0x01;
  await expect(openKeyBackup(blob, code)).rejects.toThrow(
    new KeyBackupError("the blob does not open under this code"),
  );
});

it("a tampered ciphertext does not open", async () => {
  const code = RecoveryCode.generate();
  const blob = await sealKeyBackup(randomBytes(32), code);
  blob[blob.length - 1] ^= 0x01;
  await expect(openKeyBackup(blob, code)).rejects.toThrow(
    new KeyBackupError("the blob does not open under this code"),
  );
});

it("an unsupported version and truncation are refused", async () => {
  const code = RecoveryCode.generate();
  const blob = await sealKeyBackup(randomBytes(32), code);
  const wrongVersion = blob.slice();
  wrongVersion[0] = 0x02;
  await expect(openKeyBackup(wrongVersion, code)).rejects.toThrow(
    new KeyBackupError("unsupported blob version 2"),
  );
  await expect(openKeyBackup(blob.slice(0, 29), code)).rejects.toThrow(
    new KeyBackupError("blob too short"),
  );
});

it("seal rejects malformed inputs", async () => {
  const code = RecoveryCode.generate();
  await expect(sealKeyBackup(randomBytes(16), code)).rejects.toThrow(
    new RangeError("an actor seed is 32 bytes"),
  );
  await expect(sealKeyBackup(randomBytes(32), code, randomBytes(8))).rejects.toThrow(
    new RangeError("the HKDF salt is 16 bytes"),
  );
  await expect(sealKeyBackup(randomBytes(32), code, randomBytes(16), randomBytes(8))).rejects.toThrow(
    new RangeError("the AES-GCM nonce is 12 bytes"),
  );
  expect(() => new RecoveryCode(randomBytes(8))).toThrow(new RangeError("a recovery code is 16 bytes"));
});

it("the display groups and input normalizes", () => {
  const code = RecoveryCode.generate();
  const display = code.display();
  expect(display).toMatch(/^[0-9A-HJKMNP-TV-Z]{5}(-[0-9A-HJKMNP-TV-Z]{5}){3}-[0-9A-HJKMNP-TV-Z]{6}$/);
  const sloppy = display.toLowerCase().replaceAll("1", "l").replaceAll("0", "o").replaceAll("-", " - ");
  expect(toHex(RecoveryCode.fromInput(sloppy).bytes)).toBe(toHex(code.bytes));
});

it("input refuses wrong lengths and bad characters", () => {
  expect(() => RecoveryCode.fromInput("TOO-SHORT")).toThrow(
    new RecoveryCodeLengthError("a recovery code has 26 characters"),
  );
  expect(() => RecoveryCode.fromInput("U".repeat(26))).toThrow(
    new KeyBackupError("invalid recovery-code character `U`"),
  );
  const display = RecoveryCode.generate().display();
  const nonzeroPad = display.slice(0, display.length - 1) + "Z";
  expect(() => RecoveryCode.fromInput(nonzeroPad)).toThrow(new KeyBackupError("invalid recovery code"));
});

// Only the length is a shape complaint the reader can act on; a
// full-length input that will not decode is a wrong code, and saying
// "that isn't a code" about a one-character typo misleads.
it("only a wrong length reports as a shape problem", () => {
  expect(() => RecoveryCode.fromInput("TOO-SHORT")).toThrow(RecoveryCodeLengthError);

  const display = RecoveryCode.generate().display();
  const nonzeroPad = display.slice(0, display.length - 1) + "Z";
  for (const input of [nonzeroPad, "U".repeat(26)]) {
    expect(() => RecoveryCode.fromInput(input)).toThrow(KeyBackupError);
    expect(() => RecoveryCode.fromInput(input)).not.toThrow(RecoveryCodeLengthError);
  }
});

File: android/core/crypto/src/test/kotlin/com/cogra/crypto/KeyBackupTest.kt
package com.cogra.crypto

import com.google.common.truth.Truth.assertThat
import org.junit.Assert.assertThrows
import org.junit.Test

class KeyBackupTest {

    @Test
    fun sealAndOpenRoundTrip() {
        val seed = ActorKey.generate().seed()
        val code = RecoveryCode.generate()
        val blob = sealKeyBackup(seed, code)
        assertThat(openKeyBackup(blob, code)).isEqualTo(seed)
    }

    @Test
    fun aWrongCodeDoesNotOpen() {
        val blob = sealKeyBackup(ActorKey.generate().seed(), RecoveryCode.generate())
        assertThrows(KeyBackupException::class.java) {
            openKeyBackup(blob, RecoveryCode.generate())
        }
    }

    @Test
    fun aTamperedHeaderDoesNotOpen() {
        val code = RecoveryCode.generate()
        val blob = sealKeyBackup(ActorKey.generate().seed(), code)
        // The header rides as associated data — flipping a salt bit
        // breaks the tag even though the ciphertext is untouched.
        val tampered = blob.copyOf()
        tampered[5] = (tampered[5].toInt() xor 1).toByte()
        assertThrows(KeyBackupException::class.java) { openKeyBackup(tampered, code) }
    }

    @Test
    fun aTamperedCiphertextDoesNotOpen() {
        val code = RecoveryCode.generate()
        val blob = sealKeyBackup(ActorKey.generate().seed(), code)
        val tampered = blob.copyOf()
        tampered[blob.size - 1] = (tampered[blob.size - 1].toInt() xor 1).toByte()
        assertThrows(KeyBackupException::class.java) { openKeyBackup(tampered, code) }
    }

    @Test
    fun unsupportedVersionAndTruncationAreRefused() {
        val code = RecoveryCode.generate()
        val blob = sealKeyBackup(ActorKey.generate().seed(), code)
        val wrongVersion = blob.copyOf()
        wrongVersion[0] = 0x02
        assertThrows(KeyBackupException::class.java) { openKeyBackup(wrongVersion, code) }
        assertThrows(KeyBackupException::class.java) {
            openKeyBackup(blob.copyOfRange(0, 29), code)
        }
    }

    @Test
    fun sealRejectsMalformedInputs() {
        val code = RecoveryCode.generate()
        assertThrows(IllegalArgumentException::class.java) {
            sealKeyBackup(ByteArray(16), code)
        }
        assertThrows(IllegalArgumentException::class.java) {
            sealKeyBackupWith(ByteArray(32), code, salt = ByteArray(8), nonce = ByteArray(12))
        }
        assertThrows(IllegalArgumentException::class.java) {
            sealKeyBackupWith(ByteArray(32), code, salt = ByteArray(16), nonce = ByteArray(8))
        }
        assertThrows(IllegalArgumentException::class.java) { RecoveryCode(ByteArray(8)) }
    }

    @Test
    fun displayGroupsAndInputNormalizes() {
        val code = RecoveryCode.generate()
        val display = code.display()
        assertThat(display).matches("[0-9A-HJKMNP-TV-Z]{5}(-[0-9A-HJKMNP-TV-Z]{5}){3}-[0-9A-HJKMNP-TV-Z]{6}")
        // Lowercase, confusable letters, and stray separators all read back.
        val sloppy = display.lowercase()
            .replace('1', 'l')
            .replace('0', 'o')
            .replace("-", " - ")
        assertThat(RecoveryCode.fromInput(sloppy).bytes()).isEqualTo(code.bytes())
    }

    @Test
    fun inputRefusesWrongLengthAndBadCharacters() {
        assertThrows(KeyBackupException::class.java) { RecoveryCode.fromInput("TOO-SHORT") }
        assertThrows(KeyBackupException::class.java) {
            // 'U' is not in the Crockford alphabet.
            RecoveryCode.fromInput("U".repeat(26))
        }
        // A wrong final character with nonzero pad bits is refused, not
        // silently truncated.
        val display = RecoveryCode(ByteArray(16)).display()
        assertThrows(KeyBackupException::class.java) {
            RecoveryCode.fromInput(display.dropLast(1) + "Z")
        }
    }

    // Only the length is a shape complaint the reader can act on; a
    // full-length input that will not decode is a wrong code, and
    // saying "that isn't a code" about a one-character typo misleads.
    @Test
    fun onlyAWrongLengthReportsAsAShapeProblem() {
        assertThrows(RecoveryCodeLengthException::class.java) {
            RecoveryCode.fromInput("TOO-SHORT")
        }
        val display = RecoveryCode(ByteArray(16)).display()
        for (input in listOf(display.dropLast(1) + "Z", "U".repeat(26))) {
            val thrown = assertThrows(KeyBackupException::class.java) {
                RecoveryCode.fromInput(input)
            }
            assertThat(thrown).isNotInstanceOf(RecoveryCodeLengthException::class.java)
        }
    }
}
